internal/controller/register.go

```diff
@@ -0,0 +1,66 @@
+package controller
+
+import (
+	"errors"
+	"fmt"
+
+	"k8s.io/apimachinery/pkg/types"
+	ctlr "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/client/apiutil"
+	"sigs.k8s.io/controller-runtime/pkg/manager"
+	"sigs.k8s.io/controller-runtime/pkg/predicate"
+)
+
+// NamespacedNameFilterFunc is a function that returns true if the resource should be processed by the controller.
+// If the function returns false, the controller will log the returned string.
+type NamespacedNameFilterFunc func(nsname types.NamespacedName) (bool, string)
+
+// Options contains the options for a controller.
+type Options struct {
+	// ObjectType is the type of the object the controller will watch.
+	ObjectType client.Object
+	// EventCh is the channel where the controller will send events (from the events package).
+	EventCh chan<- interface{}
```

Comment:
This is a shared channel across all the controllers, right?

Does it make sense to separate this field from the controller `Options` so we don't have to provide the same channel every time we register a controller?

Reply:
I removed eventCh from Options.

However, because not all Options are used by ReconcilierImpl, I refactored ReconcilierImpl to have those options as fields and the corresponding arguments in the NewReconcilerImpl --  I didn't want to have another options struct.  Part of the reason is that I don't think we should have that type exported at all .`ReconcilerImpl` it is only used inside the controllers package. The reason I made it exported is that we agreed we will use ginkgo to test public interfaces only. However, perhaps we can relax that agreement?

Reply:
another idea: moving the registration logic (RegisterController) to the manager package (since we're calling manager method directly in that function, I think it even makes more sense) and leave ReconcilerImpl exported.

Reply:
> The reason I made it exported is that we agreed we will use ginkgo to test public interfaces only. However, perhaps we can relax that agreement?

If we can write better unit tests using ginkgo & gomega, then I think we should do it -- regardless of whether the interface is public. I don't think we should be making design decisions based on a self-imposed tooling restriction. This might be a good time to revisit our testing agreement/strategy. 

> another idea: moving the registration logic (RegisterController) to the manager package (since we're calling manager method directly in that function, I think it even makes more sense) and leave ReconcilerImpl exported.

I like this idea.

Reply:
Refactored:
- rename controller package to reconciler
- move controller registration to manager package
- refactor registerControllers function - now it takes controller configs
as an argument